confgenerator/confgenerator.go

```diff
@@ -84,6 +84,9 @@ func (uc *UnifiedConfig) GenerateOtelConfig(hostInfo *host.InfoStat) (string, er
 					"instrumentation_library_labels": false,
 					// Omit service labels, which break agent metrics.
 					"service_resource_labels": false,
+					"resource_filters": []map[string]interface{}{
+						{"prefix": "sqlserver.database.name"},
+					},
```

Comment:
Why are we adding this across the board? Is there really no way to confine this change to the sqlserver receiver? Does this not break existing metrics?

Reply:
@igorpeshansky that's a good call out that I am aware of. I could minimize it a bit possibly by only including that filter if the sql server pipeline exists....but it still would apply to everything being exported in that circumstance because there is only a single google exporter for all apps (as far as I know).

Reply:
To add a little more to this, the work that was done to the google exporter (to be able to handle resource attributes) allows it to be configured with filters in order to translate the OTEL resource attributes to labels. Outside of this, nothing was added to account for the issues of OTEL receivers with resource attributes.

So unless I'm missing something, this is the best way I could think of dealing with the issue without coming up with a significant change to how the ops agent handles pipelines (which I know is not great because this change is on a global exporter which affects all pipelines...hopefully mitigated by good resource attribute naming).

Reply:
@StefanKurek I may be missing something about how metrics are modified, but it looks like these resource fields would be added indiscriminately. If not, a comment with a pointer to why this works would be really helpful.

@jefferbrecht I suspect we might just not be running the expected metrics tests with the `sqlserver` pipeline present. One way to be absolutely sure is to build a prototype package from this branch, install it, enable the `sqlserver` receiver, and see if it starts getting API failures due to extraneous labels. I can try to do that tomorrow unless someone beats me to it.

For the record, the substitution of `.` to `_` happens [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/main/exporter/metric/metric.go#L641-L643).

Reply:
@igorpeshansky I believe this is the area of code you're looking for https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/836e6fe9832c9f19146dabab29a8b55d9511459f/exporter/collector/monitoredresource.go#L201

Reply:
IIUC the labels are only added if the prefix matches the attribute key. It will be added indiscriminately in the sense that it's not receiver-specific, but effectively this is the only receiver whose attribute keys contain that prefix. There's some documentation [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/3a2c57209765f4752bd140a007612a81f58489c1/exporter/collector/config.go#L92-L94), and the code for matching/adding is [here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/836e6fe9832c9f19146dabab29a8b55d9511459f/exporter/collector/monitoredresource.go#L201).

The `sqlserver` tests were running correctly last I checked:

```
2022/04/22 17:49:32 nonEmptySeries() iterator supplied err <nil> and series metric:{type:"workload.googleapis.com/sqlserver.user.connection.count"}  resource:{type:"gce_instance"  labels:{key:"instance_id"  value:"9188105535274022384"}  labels:{key:"project_id"  value:"stackdriver-test-143416"}  labels:{key:"zone"  value:"us-central1-b"}}  metric_kind:GAUGE  value_type:DOUBLE  points:{interval:{end_time:{seconds:1650674969  nanos:104023000}  start_time:{seconds:1650674969  nanos:104023000}}  value:{double_value:0}}
2022/04/22 17:49:32 Found representative metric, sleeping before checking remaining metrics
...
2022/04/22 17:50:42 Skipping optional or representative metric workload.googleapis.com/sqlserver.user.connection.count
2022/04/22 17:50:42 nonEmptySeries() iterator supplied err <nil> and series metric:{type:"workload.googleapis.com/sqlserver.transaction_log.flush.data.rate"  labels:{key:"sqlserver_database_name"  value:"master"}}  resource:{type:"gce_instance"  labels:{key:"instance_id"  value:"9188105535274022384"}  labels:{key:"project_id"  value:"stackdriver-test-143416"}  labels:{key:"zone"  value:"us-central1-b"}}  metric_kind:GAUGE  value_type:DOUBLE  points:{interval:{end_time:{seconds:1650675029  nanos:74760000}  start_time:{seconds:1650675029  nanos:74760000}}  value:{double_value:0}}  points:{interval:{end_time:{seconds:1650674969  nanos:103151000}  start_time:{seconds:1650674969  nanos:103151000}}  value:{double_value:0}}
```

Or are you saying that if _any_ attribute key matches that prefix, then _all_ metrics passing through that exporter, regardless of _their_ attributes, might _also_ get labels added? If that's what you're saying then I assume we can just check the agent metrics from the above test run and confirm that they don't have extraneous labels, since agent metrics pass through the same exporter:

```
{
  "timeSeries": [
    {
      "metric": {
        "labels": {
          "version": "google-cloud-ops-agent-logging/2.14.0-windows-ltsc2019"
        },
        "type": "agent.googleapis.com/agent/uptime"
      },
      "resource": {
        "type": "gce_instance",
        "labels": {
          "instance_id": "9188105535274022384",
          "zone": "us-central1-b",
          "project_id": "stackdriver-test-143416"
        }
      },
      ...
```

But you're right that the tests don't check for this currently.